Log in rust

selaux opened the pull request:
This is the opposite of #832. Here we replace SLOG by a Logger implemented in Rust. Felt way more cleaner to implement that #832. One disadvantage might be that logging is slower due to conversion overhead and C++ doing `sprintf`s regardless of whether the log message is used in the end.

A few differences to before:
- Log format has changed
- Instead of SLOG topics filenames are used for C++ and module paths for Rust code

Changes to LOG format:

**Before**:
```
2019/06/28 18:22:29 DEBUG   [SGP             ] Initializing Memory Manager
2019/06/28 18:22:29 DEBUG   [SGP             ] Initializing Game Resources
2019/06/28 18:22:29 INFO    [SGP             ] ------------------------------------------------------------------------------
2019/06/28 18:22:29 INFO    [SGP             ] JA2 Home Dir:                  '/home/stefan/.ja2'
2019/06/28 18:22:29 INFO    [SGP             ] Root game resources directory: '/home/stefan/.ja2/orig_data'
2019/06/28 18:22:29 INFO    [SGP             ] Extra data directory:          '/home/stefan/workspace/ja2-stracciatella/_bin/.'
2019/06/28 18:22:29 INFO    [SGP             ] Data directory:                '/home/stefan/.ja2/orig_data/data'
2019/06/28 18:22:29 INFO    [SGP             ] Tilecache directory:           '/home/stefan/.ja2/orig_data/data/Tilecache'
2019/06/28 18:22:29 INFO    [SGP             ] Saved games directory:         '/home/stefan/.ja2/SavedGames'
2019/06/28 18:22:29 INFO    [SGP             ] ------------------------------------------------------------------------------
2019/06/28 18:22:30 DEBUG   [Default CM      ] Loaded calibre 0 NOAMMO
```

**After**:
```
16:26:58 [DEBUG] sgp/SGP.cc: Initializing Memory Manager
16:26:58 [DEBUG] sgp/SGP.cc: Initializing Game Resources
16:26:58 [ INFO] sgp/SGP.cc: ------------------------------------------------------------------------------
16:26:58 [ INFO] sgp/SGP.cc: JA2 Home Dir:                  '/home/stefan/.ja2'
16:26:58 [ INFO] sgp/SGP.cc: Root game resources directory: '/home/stefan/.ja2/orig_data'
16:26:58 [ INFO] sgp/SGP.cc: Extra data directory:          '/home/stefan/workspace/ja2-stracciatella/_bin/.'
16:26:58 [ INFO] sgp/SGP.cc: Data directory:                '/home/stefan/.ja2/orig_data/data'
16:26:58 [ INFO] sgp/SGP.cc: Tilecache directory:           '/home/stefan/.ja2/orig_data/data/Tilecache'
16:26:58 [ INFO] sgp/SGP.cc: Saved games directory:         '/home/stefan/.ja2/SavedGames'
16:26:58 [ INFO] sgp/SGP.cc: ------------------------------------------------------------------------------
16:26:58 [DEBUG] externalized/DefaultContentManager.cc: Loaded calibre 0 NOAMMO

```

lynxlynxlynx commented on rust/src/config/cli.rs:
```diff
@@ -116,6 +117,10 @@ impl Cli {
+                if m.opt_str("datadir").is_some() {
+                    warn!("The `datadir` command line argument is deprecated, use `gamedir` instead");
+                }
+
```
is this intentional? Appears to be for/from the PR about the argument renaming.

selaux replied:
Yes, I just wanted to have an easy usecase for logging in Rust. And those were not logged yet.

selaux commented:
Please let me know if this is something we want, then I will finish this one and close the other PR.

lynxlynxlynx commented:
Is there any tangible speed difference or is it more like noise?

flaviojs commented on src/launcher/main.cc:
```diff
@@ -1,13 +1,11 @@
 int main(int argc, char* argv[]) {
-	SLOG_Init(SLOG_STDERR, "stracciatella-launcher.log");
-	SLOG_SetLevel(SLOG_WARNING, SLOG_WARNING);
+	Logger_Init("stracciatella-launcher.log");
```
I would use the exe name `ja2-launcher.log` to match `ja2.log`.
Or change `ja2.log` to `stracciatella.log` to match this.

flaviojs commented:
Tried it and it didn't feel slower, so the speed difference is on the level of noise.

The new format only shows the time.
There should be at least one line with the full date to use as reference.
I would add it when the logger is initialized.

flaviojs commented on rust/src/logger.rs:
```diff
@@ -0,0 +1,152 @@
+
+use lazy_static::lazy_static;
+use log::*;
+use simplelog::*;
```
It's better to avoid `*` imports unless it's something like a prelude. (`c::common` is like a prelude)
It's hard to see which crate log stuff comes from like this.
There is also potential for conflict when a new symbol is added to one of those crates.

flaviojs commented on rust/src/logger.rs:
```diff
@@ -0,0 +1,152 @@
+
+lazy_static! {
+    /// Global log level used to filter messages at runtime
+    static ref GLOBAL_LOG_LEVEL: Arc<atomic::AtomicUsize> = {
```
You're using AtomicUsize so you don't need `lazy_static` or 'Arc`.
See "Atomic Access" in https://docs.rust-embedded.org/book/concurrency/

selaux replied:
True, I used this so I can call `LogLevel::default`, but I think just initializing with a specific value makes it clear which log level is default as well...

flaviojs commented on rust/src/logger.rs:
```diff
@@ -0,0 +1,152 @@
+#[derive(Debug, PartialEq, Copy, Clone)]
+#[repr(C)]
+/// Enum to represent log levels in the application
+pub enum LogLevel {
```
This appears to be for C code use.
If that is the case then it should be moved to `c::logger`.
`log::Level` can be converted to `usize` and back too (for `GLOBAL_LOG_LEVEL`).

selaux replied:
Well it can be used in rust, e.g. in `set_level`, so I think this should stay here. Having `set_level` getting a `usize` argument seems confusing to me. The `#[repr(C)]` is only there to make it possible to reexport this inside `c::logger`.

flaviojs replied:
It's fine like this as well.

What I meant was `Logger::set_level` receiving a `log::Level` and store it as a `usize` in  `GLOBAL_LOG_LEVEL`.
Then `RuntimeLevelFilter::get_global_log_level` would convert it back from a `usize`.
This way `LogLevel` only interfaces with C code (mapping to internal `log::Level` values) so it would be a part of `c::logger`.

flaviojs commented on rust/src/logger.rs:
```diff
@@ -0,0 +1,152 @@
+    ///
+    /// Needs to be called once at start of the game engine. Any log messages send
+    /// before will be discarded.
+    pub fn init(log_file: &Path) {
```
After initializing, write a log line with the current date?
That way the other log lines, which only show time, have a fixed reference point.

selaux replied:
I just added the date to log entries.

flaviojs commented on src/sgp/Logger.cc:
```diff
@@ -0,0 +1,30 @@
+  #ifdef ENABLE_ASSERTS
+    if (tag == DEBUG_TAG_ASSERTS)
+    {
+      abort();
```
The `abort()` can be moved to the Assert defines. This errors because `tag` is no longer there.

selaux commented:
Addressed the comments, thanks for the review. Log format now includes the date as well:

```
2019-08-08T18:50:52 [ INFO] sgp/SGP.cc: ------------------------------------------------------------------------------
2019-08-08T18:50:52 [ INFO] sgp/SGP.cc: JA2 Home Dir:                  '/home/stefan/.ja2'
2019-08-08T18:50:52 [ INFO] sgp/SGP.cc: Root game resources directory: '/home/stefan/.ja2/orig_data'
2019-08-08T18:50:52 [ INFO] sgp/SGP.cc: Extra data directory:          '/home/stefan/workspace/ja2-stracciatella/_bin/.'
2019-08-08T18:50:52 [ INFO] sgp/SGP.cc: Data directory:                '/home/stefan/.ja2/orig_data/data'
2019-08-08T18:50:52 [ INFO] sgp/SGP.cc: Tilecache directory:           '/home/stefan/.ja2/orig_data/data/Tilecache'
2019-08-08T18:50:52 [ INFO] sgp/SGP.cc: Saved games directory:         '/home/stefan/.ja2/SavedGames'
2019-08-08T18:50:52 [ INFO] sgp/SGP.cc: ------------------------------------------------------------------------------
2019-08-08T18:50:53 [ INFO] game/GameLoop.cc: Version Label: Stracciatella "v0.16.x"
2019-08-08T18:50:53 [ INFO] game/GameLoop.cc: Version #:     Build 04.12.02
```

flaviojs commented:
LGTM
